lib/perl/Genome/Model/Tools/Analysis/Coverage/CoveragePlot.R

```diff
@@ -0,0 +1,1549 @@
+{
+    # returns a vector containing the minimum and maximum of all the given arguments in the given order
+    s <- cs_range(scale);
+    
+    # extends a numerical range by a small percentage, i.e., fraction, on both sides
+    r <- extendrange(scale, f=factor);
+    
+    if (s[3]) return(c(r[2], r[1]))
+    else return(c(r[1], r[2]));
+
+    # updated on 2010/09/22
+    #r <- range(scale);
+    #return (range(scale) + c(-factor, factor) * diff(range(scale)));
+}
+
+
+
+cs_range <- function(x)
+{
+    # gets the minimum and maximum
+    r <- range(x);
+    
+    if (x[1] <= x[length(x)]) reverse <- FALSE
+    else reverse <- TRUE;
+    
+    if (reverse) return(c(r[2], r[1], reverse))
+    else return(c(r[1], r[2], reverse));
+}
+
+
+
+cs_extend_scale <- function(scale, range=NULL)
+{
+    if(is.null(range)) return(scale);
+    
+    # extends a numerical range by a small percentage, i.e., fraction, on both sides
+    breaks <- length(scale);
+    e <- extendrange(scale, r=range(range, na.rm=T), f=0)
+
+    # gets the data scope
+    s <- cs_range(e);
+        
+    # gets pretty breakpoints
+    p <- pretty(e, n=breaks);
+    
+    if (s[3]) return(rev(p))
+    else return(p);
+}
+
+
+
+cs_range_limit2 <- function(x, y, xlim=NULL, ylim=NULL)
+{
+    if (is.null(xlim)) stop("xlim not defined");
+    
+    # checks the data region
+    minx = min(x);
+    maxx = max(x);
+    miny = min(y);
+    maxy = max(y);
+    
+    #print(c(minx, maxx, miny, maxy));   # for debug
+    #print(xlim);  # for debug
+    #print(ylim);  # for debug
+        
+    # if all x and y data points are within the given range
+    if (xlim[1] <= minx & maxx <= xlim[2])
+    {
+        if (is.null(ylim))
+        {
+            return(list(x=x, y=y));
+        }
+        else if (ylim[1] <= miny & maxy <= ylim[2])
+        {
+            return(list(x=x, y=y));
+        }
+    }
+    
+    # selects the data points that are within the given range
+    data <- list(x=vector(mode="numeric"), y=vector(mode="numeric"));
+    
+    # adds into a new dataset
+    idx <- 1;
+    for(i in 1:length(x))
+    {
+        if (is.null(ylim))
+        {
+            if (xlim[1] <= x[i] & x[i] <= xlim[2])
+            {
+                # a new dataset
+                data$x[idx] <- x[i];
+                data$y[idx] <- y[i];
+                
+                # increases the index number
+                idx <- idx + 1;
+            }
+        }
+        else
+        {
+            if (xlim[1] <= x[i] & x[i] <= xlim[2] & ylim[1] <= y[i] & y[i] <= ylim[2])
+            {
+                # a new dataset
+                data$x[idx] <- x[i];
+                data$y[idx] <- y[i];
+                
+                # increases the index number
+                idx <- idx + 1;
+            }
+        }
+    }
+    
+    #print(data);    # for debug
+    
+    return(data);
+}
+
+
+
+cs_range_limit3 <- function(x, y, z, xlim=NULL, ylim=NULL, zlim=NULL)

```

Comment:
Many of these function names could be made more descriptive. The difference between cs_range_limit2 and cs_range_limit3 is not at all apparent from the function name.